File: py/idkollen_client/models/mitid.py
from __future__ import annotations

from typing import Annotated, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from idkollen_client.models.common import ApiErrorCode


class _Base(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class MitIdAuthRequest(_Base):
    redirect_url: Optional[str] = None
    reference_text: Optional[str] = None
    request_phone: Optional[bool] = None
    request_email: Optional[bool] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None


class MitIdBackchannelAuthRequest(_Base):
    ssn: str
    binding_message: str
    callback_url: Optional[str] = None
    ref_id: Optional[str] = None


class MitIdSignRequest(_Base):
    text: str
    redirect_url: Optional[str] = None
    ref_id: Optional[str] = None


class MitIdPending(_Base):
    status: Literal["PENDING"]
    id: str
    ref_id: Optional[str] = None
    url: Optional[str] = None
    binding_message: Optional[str] = None


class MitIdSignResult(_Base):
    checksum: str


class MitIdCompleted(_Base):
    status: Literal["COMPLETED"]
    id: str
    ref_id: Optional[str] = None
    ssn: str
    name: str
    given_name: str
    surname: str
    phone: Optional[str] = None
    email: Optional[str] = None
    address: Optional[str] = None
    birth_date: Optional[str] = None
    pid: Optional[str] = None
    bank_id: Optional[str] = None
    sign_result: Optional[MitIdSignResult] = None


class MitIdFailed(_Base):
    status: Literal["FAILED"]
    id: str
    ref_id: Optional[str] = None
    error: ApiErrorCode


MitIdStatus = Annotated[
    MitIdPending | MitIdCompleted | MitIdFailed,
    Field(discriminator="status"),
]

File: csharp/src/Idkollen.Client/Models/MitId.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Idkollen.Client.Models;

public sealed record MitIdAuthRequest
{
    public string? RedirectUrl { get; init; }
    public string? ReferenceText { get; init; }
    public bool? RequestPhone { get; init; }
    public bool? RequestEmail { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
}

public sealed record MitIdBackchannelAuthRequest(string Ssn, string BindingMessage)
{
    public string? CallbackUrl { get; init; }
    public string? RefId { get; init; }
}

public sealed record MitIdSignRequest(string Text)
{
    public string? RedirectUrl { get; init; }
    public string? RefId { get; init; }
}

public sealed record MitIdSignResult(string Checksum);

public interface IMitIdStatus { }

public sealed record MitIdPending(
    string Id,
    string? RefId,
    string? Url,
    string? BindingMessage
) : IMitIdStatus;

public sealed record MitIdCompleted(
    string Id,
    string? RefId,
    string Ssn,
    string Name,
    string GivenName,
    string Surname,
    string? Phone,
    string? Email,
    string? Address,
    string? BirthDate,
    string? Pid,
    string? BankId,
    MitIdSignResult? SignResult
) : IMitIdStatus;

public sealed record MitIdFailed(string Id, string? RefId, string Error) : IMitIdStatus;

internal sealed class MitIdStatusConverter : JsonConverter<IMitIdStatus>
{
    public override IMitIdStatus Read(ref Utf8JsonReader reader, Type _, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        var status = doc.RootElement.GetProperty("status").GetString();
        var raw = doc.RootElement.GetRawText();
        return status switch
        {
            "PENDING" => JsonSerializer.Deserialize<MitIdPending>(raw, options)!,
            "COMPLETED" => JsonSerializer.Deserialize<MitIdCompleted>(raw, options)!,
            "FAILED" => JsonSerializer.Deserialize<MitIdFailed>(raw, options)!,
            _ => throw new JsonException($"Unknown mitid status: {status}"),
        };
    }

    public override void Write(Utf8JsonWriter w, IMitIdStatus v, JsonSerializerOptions o)
        => throw new NotSupportedException();
}
